// @volar/vue-language-server/out/protocol.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParseSFCRequest = exports.GetConvertAttrCasingEditsRequest = exports.GetConvertTagCasingEditsRequest = exports.DetectNameCasingRequest = exports.GetComponentMeta = exports.GetVueCompilerOptionsRequest = void 0;
const vscode = require("vscode-languageserver-protocol");
var GetVueCompilerOptionsRequest;
(function (GetVueCompilerOptionsRequest) {
    GetVueCompilerOptionsRequest.type = new vscode.RequestType('volar/vueCompilerOptions');
})(GetVueCompilerOptionsRequest = exports.GetVueCompilerOptionsRequest || (exports.GetVueCompilerOptionsRequest = {}));
var GetComponentMeta;
(function (GetComponentMeta) {
    GetComponentMeta.type = new vscode.RequestType('volar/meta');
})(GetComponentMeta = exports.GetComponentMeta || (exports.GetComponentMeta = {}));
var DetectNameCasingRequest;
(function (DetectNameCasingRequest) {
    DetectNameCasingRequest.type = new vscode.RequestType('volar/detectTagCasing');
})(DetectNameCasingRequest = exports.DetectNameCasingRequest || (exports.DetectNameCasingRequest = {}));
var GetConvertTagCasingEditsRequest;
(function (GetConvertTagCasingEditsRequest) {
    GetConvertTagCasingEditsRequest.type = new vscode.RequestType('volar/convertTagNameCasing');
})(GetConvertTagCasingEditsRequest = exports.GetConvertTagCasingEditsRequest || (exports.GetConvertTagCasingEditsRequest = {}));
var GetConvertAttrCasingEditsRequest;
(function (GetConvertAttrCasingEditsRequest) {
    GetConvertAttrCasingEditsRequest.type = new vscode.RequestType('volar/convertAttrNameCasing');
})(GetConvertAttrCasingEditsRequest = exports.GetConvertAttrCasingEditsRequest || (exports.GetConvertAttrCasingEditsRequest = {}));
var ParseSFCRequest;
(function (ParseSFCRequest) {
    ParseSFCRequest.type = new vscode.RequestType('volar/parseSfc');
})(ParseSFCRequest = exports.ParseSFCRequest || (exports.ParseSFCRequest = {}));
