// @volar-plugins/pug/out/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const pug = require("@volar/pug-language-service");
const html_1 = require("@volar-plugins/html");
function default_1() {
    const htmlPlugin = (0, html_1.default)({});
    const pugDocuments = new WeakMap();
    let context;
    let pugLs;
    return Object.assign(Object.assign({}, htmlPlugin), { getPugLs: () => pugLs, getPugDocument,
        setup(_context) {
            var _a;
            (_a = htmlPlugin.setup) === null || _a === void 0 ? void 0 : _a.call(htmlPlugin, _context);
            pugLs = pug.getLanguageService(htmlPlugin.getHtmlLs());
            context = _context;
        }, complete: {
            on(document, position, _) {
                return worker(document, (pugDocument) => {
                    if (!context.env.documentContext)
                        return;
                    return pugLs.doComplete(pugDocument, position, context.env.documentContext);
                });
            },
        }, validation: {
            onSyntactic(document) {
                return worker(document, (pugDocument) => {
                    if (pugDocument.error) {
                        return [{
                                code: pugDocument.error.code,
                                message: pugDocument.error.msg,
                                range: {
                                    start: { line: pugDocument.error.line, character: pugDocument.error.column },
                                    end: { line: pugDocument.error.line, character: pugDocument.error.column },
                                },
                            }];
                    }
                });
            },
        }, doHover(document, position) {
            return worker(document, (pugDocument) => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const hoverSettings = yield ((_a = context.env.configurationHost) === null || _a === void 0 ? void 0 : _a.getConfiguration('html.hover'));
                return pugLs.doHover(pugDocument, position, hoverSettings);
            }));
        },
        findDocumentHighlights(document, position) {
            return worker(document, (pugDocument) => {
                return pugLs.findDocumentHighlights(pugDocument, position);
            });
        },
        findDocumentLinks(document) {
            return worker(document, (pugDocument) => {
                if (!context.env.documentContext)
                    return;
                return pugLs.findDocumentLinks(pugDocument, context.env.documentContext);
            });
        },
        findDocumentSymbols(document) {
            return worker(document, (pugDocument) => {
                return pugLs.findDocumentSymbols(pugDocument);
            });
        },
        getFoldingRanges(document) {
            return worker(document, (pugDocument) => {
                return pugLs.getFoldingRanges(pugDocument);
            });
        },
        getSelectionRanges(document, positions) {
            return worker(document, (pugDocument) => {
                return pugLs.getSelectionRanges(pugDocument, positions);
            });
        },
        doAutoInsert(document, position, insertContext) {
            return __awaiter(this, void 0, void 0, function* () {
                return worker(document, (pugDocument) => __awaiter(this, void 0, void 0, function* () {
                    var _a, _b, _c;
                    const lastCharacter = insertContext.lastChange.text[insertContext.lastChange.text.length - 1];
                    if (insertContext.lastChange.rangeLength === 0 && lastCharacter === '=') {
                        const enabled = (_b = (yield ((_a = context.env.configurationHost) === null || _a === void 0 ? void 0 : _a.getConfiguration('html.autoCreateQuotes')))) !== null && _b !== void 0 ? _b : true;
                        if (enabled) {
                            const text = pugLs.doQuoteComplete(pugDocument, position, yield ((_c = context.env.configurationHost) === null || _c === void 0 ? void 0 : _c.getConfiguration('html.completion')));
                            if (text) {
                                return text;
                            }
                        }
                    }
                }));
            });
        } });
    function worker(document, callback) {
        const pugDocument = getPugDocument(document);
        if (!pugDocument)
            return;
        return callback(pugDocument);
    }
    function getPugDocument(document) {
        if (document.languageId !== 'jade')
            return;
        const cache = pugDocuments.get(document);
        if (cache) {
            const [cacheVersion, cacheDoc] = cache;
            if (cacheVersion === document.version) {
                return cacheDoc;
            }
        }
        const doc = pugLs.parsePugDocument(document.getText());
        pugDocuments.set(document, [document.version, doc]);
        return doc;
    }
}
exports.default = default_1;
